framework/include/constraints/NodeElemConstraint.h

```diff
@@ -0,0 +1,280 @@
+ */
+class NodeElemConstraint : public Constraint,
+                           public NeighborCoupleableMooseVariableDependencyIntermediateInterface,
+                           public NeighborMooseVariableInterface<Real>
+{
+public:
+  NodeElemConstraint(const InputParameters & parameters);
+  virtual ~NodeElemConstraint();
+
+  /**
+   * Compute the value the slave node should have at the beginning of a timestep.
+   */
+  void computeSlaveValue(NumericVector<Number> & current_solution);
+
+  /**
+   * Computes the residual Nodal residual.
+   */
+  virtual void computeResidual();
+
+  /**
+   * Computes the jacobian for the current element.
+   */
+  virtual void computeJacobian();
+
+  /**
+   * Computes d-residual / d-jvar...
+   */
+  virtual void computeOffDiagJacobian(unsigned int jvar);
+
+  /**
+   * Gets the indices for all dofs connected to the constraint
+   */
+  virtual void getConnectedDofIndices(unsigned int var_num);
+
+  /**
+   * Whether or not this constraint should be applied.
+   *
+   * Get's called once per slave node per variable.
+   */
+  virtual bool shouldApply() { return true; }
+
+  /**
+   * Compute reaction force of the constraint depending on the formulation
+   */
+  virtual void computeReactionForce(){};
+
+  /**
+   * Whether or not the slave's residual should be overwritten.
+   *
+   * When this returns true the slave's residual as computed by the constraint will _replace_
+   * the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSlaveResidual();
+
+  /**
+   * Whether or not the slave's Jacobian row should be overwritten.
+   *
+   * When this returns true the slave's Jacobian row as computed by the constraint will _replace_
+   * the residual previously at that node for that variable.
+   */
+  virtual bool overwriteSlaveJacobian() { return overwriteSlaveResidual(); };
+
+  /**
+   * The variable on the Master side of the domain.
+   */
+  virtual MooseVariable & masterVariable() { return _master_var; }
+
+  // TODO: Make this protected or add an accessor
+  // Do the same for all the other public members
+  SparseMatrix<Number> * _jacobian;
+
+protected:
+  /**
+   * Compute the value the slave node should have at the beginning of a timestep.
+   */
+  virtual Real computeQpSlaveValue() = 0;
+
+  /**
+   * This is the virtual that derived classes should override for computing the residual.
+   */
+  virtual Real computeQpResidual(Moose::ConstraintType type) = 0;
+
+  /**
+   * This is the virtual that derived classes should override for computing the Jacobian.
+   */
+  virtual Real computeQpJacobian(Moose::ConstraintJacobianType type) = 0;
+
+  /**
+   * This is the virtual that derived classes should override for computing the off-diag Jacobian.
+   */
+  virtual Real computeQpOffDiagJacobian(Moose::ConstraintJacobianType /*type*/,
+                                        unsigned int /*jvar*/)
+  {
+    return 0;
+  }
+
+  /// coupling interface:
```

Comment:
Do we really need all of these `coupled*()` methods?  I question whether we need them, but since I guess a variety of constraints could derive from this and actually need them, there's an argument for keeping them.